test_runner/batch_others/test_timeline_size.py

```diff
@@ -176,3 +178,116 @@ def test_timeline_size_quota(neon_env_builder: NeonEnvBuilder):
+
+    with closing(pg.connect()) as conn:
+        with conn.cursor() as cur:
+            cur.execute("CREATE TABLE foo (t text)")
+            cur.execute("""
+                INSERT INTO foo
+                    SELECT 'long string to consume some space' || g
+                    FROM generate_series(1, 100000) g
+            """)
+
+    env.pageserver.safe_psql(f"checkpoint {env.initial_tenant.hex} {new_timeline_id.hex}")
+    env.pageserver.safe_psql(f"compact {env.initial_tenant.hex} {new_timeline_id.hex}")
+    assert_physical_size(env, env.initial_tenant, new_timeline_id)
+
+
+def test_timeline_physical_size_post_gc(neon_env_builder: NeonEnvBuilder):
+    # Disable background compaction as we don't want it to happen after `get_physical_size` request
+    # and before checking the expected size on disk, which makes the assertion failed
+    neon_env_builder.pageserver_config_override = "tenant_config={checkpoint_distance=10000, compaction_period='10m', pitr_interval='1s'}"
+
+    env = neon_env_builder.init_start()
+
+    new_timeline_id = env.neon_cli.create_branch('test_timeline_physical_size_post_compaction')
+    pg = env.postgres.create_start("test_timeline_physical_size_post_compaction")
+
+    with closing(pg.connect()) as conn:
+        with conn.cursor() as cur:
+            cur.execute("CREATE TABLE foo (t text)")
+            cur.execute("""
+                INSERT INTO foo
+                    SELECT 'long string to consume some space' || g
+                    FROM generate_series(1, 100000) g
+            """)
+            env.pageserver.safe_psql(f"checkpoint {env.initial_tenant.hex} {new_timeline_id.hex}")
+            cur.execute("""
+                INSERT INTO foo
+                    SELECT 'long string to consume some space' || g
+                    FROM generate_series(1, 100000) g
+            """)
+            env.pageserver.safe_psql(f"checkpoint {env.initial_tenant.hex} {new_timeline_id.hex}")
+
+    env.pageserver.safe_psql(f"do_gc {env.initial_tenant.hex} {new_timeline_id.hex} 0")
+    assert_physical_size(env, env.initial_tenant, new_timeline_id)
+
+
+def assert_physical_size(env: NeonEnv, tenant_id: UUID, timeline_id: UUID):
+    """Check the current physical size returned from timeline API
+    matches the total physical size of the timeline on disk"""
+    client = env.pageserver.http_client()
+    res = assert_timeline_local(client, tenant_id, timeline_id)
+    timeline_path = f"{env.repo_dir}/tenants/{tenant_id.hex}/timelines/{timeline_id.hex}/"
+    assert res["local"]["current_physical_size"] == get_timeline_dir_size(timeline_path)
+
+
+def get_timeline_dir_size(path: str) -> int:
+    """Get the timeline directory's total size, which only counts the layer files' size."""
+    sz = 0
+    for f in os.listdir(path):
+        f_path = os.path.join(path, f)
+        # "simple" check if a file is a layer file
+        if os.path.isfile(f_path) and (not f.endswith('temp')) and ("__" in f):
```

Comment:
Should we explicitly specify which kind of files we ignore? Or other way around specify that we count only layer files :) E g s3 download code creates temporary files with extension `temp_download` and it will go through this filter. Maybe entirely avoid files with extensions?

Reply:
Yeah, we should. The current version is good enough for the new tests but I guess it's still better to be more correct on this, so future tests can use this function.

Reply:
Updated in https://github.com/neondatabase/neon/pull/2126/commits/2cdc14709411f1eb077426c2be596ac7ce4d914c. Now it filters by parsing the delta files, similar to what we do when loading the layer map.